presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/DetermineSemiJoinDistributionType.java

```diff
@@ -0,0 +1,162 @@
+package com.facebook.presto.sql.planner.iterative.rule;
+
+import com.facebook.presto.cost.CostComparator;
+import com.facebook.presto.cost.PlanNodeCostEstimate;
+import com.facebook.presto.cost.PlanNodeStatsEstimate;
+import com.facebook.presto.cost.StatsProvider;
+import com.facebook.presto.cost.TaskCountEstimator;
+import com.facebook.presto.matching.Captures;
+import com.facebook.presto.matching.Pattern;
+import com.facebook.presto.sql.analyzer.FeaturesConfig.JoinDistributionType;
+import com.facebook.presto.sql.planner.TypeProvider;
+import com.facebook.presto.sql.planner.iterative.Rule;
+import com.facebook.presto.sql.planner.plan.PlanNode;
+import com.facebook.presto.sql.planner.plan.SemiJoinNode;
+import com.google.common.collect.Ordering;
+import io.airlift.units.DataSize;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Optional;
+
+import static com.facebook.presto.SystemSessionProperties.getJoinDistributionType;
+import static com.facebook.presto.SystemSessionProperties.getJoinMaxBroadcastTableSize;
+import static com.facebook.presto.cost.CostCalculatorWithEstimatedExchanges.calculateJoinCostWithoutOutput;
+import static com.facebook.presto.sql.analyzer.FeaturesConfig.JoinDistributionType.AUTOMATIC;
+import static com.facebook.presto.sql.planner.plan.Patterns.semiJoin;
+import static com.facebook.presto.sql.planner.plan.SemiJoinNode.DistributionType.PARTITIONED;
+import static com.facebook.presto.sql.planner.plan.SemiJoinNode.DistributionType.REPLICATED;
+import static java.util.Objects.requireNonNull;
+
+/**
+ * This rule must run after the distribution type has already been set for delete queries,
+ * since semi joins in delete queries must be replicated.
+ * Once we have better pattern matching, we can fold that optimizer into this one.
+ */
+public class DetermineSemiJoinDistributionType
+        implements Rule<SemiJoinNode>
+{
+    private final TaskCountEstimator taskCountEstimator;
+    private final CostComparator costComparator;
+
+    private static final Pattern<SemiJoinNode> PATTERN = semiJoin().matching(semiJoin -> !semiJoin.getDistributionType().isPresent());
+
+    public DetermineSemiJoinDistributionType(CostComparator costComparator, TaskCountEstimator taskCountEstimator)
+    {
+        this.costComparator = requireNonNull(costComparator, "costComparator is null");
+        this.taskCountEstimator = requireNonNull(taskCountEstimator, "taskCountEstimator is null");
+    }
+
+    @Override
+    public Pattern<SemiJoinNode> getPattern()
+    {
+        return PATTERN;
+    }
+
+    @Override
+    public Result apply(SemiJoinNode semiJoinNode, Captures captures, Context context)
+    {
+        JoinDistributionType joinDistributionType = getJoinDistributionType(context.getSession());
+        if (joinDistributionType == AUTOMATIC) {
+            return Result.ofPlanNode(getCostBasedDistributionType(semiJoinNode, context));
+        }
+
+        else if (joinDistributionType.canPartition()) {
```

Comment:
nit: `joinDistributionType.canPartition()` looks confusing. It looks like it does something more that simply checking the enum. Thus i would suggest replace these `ifs` with a switch.